zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/UserTaskListenersTest.java

```diff
@@ -103,4 +92,71 @@ void shouldCompleteUserTaskWithCompleteTaskListener() {
+    final var userTaskKey =
+        resourcesHelper.createSingleUserTask(
+            task ->
+                task.zeebeTaskListener(
+                    listener ->
+                        listener
+                            .complete()
+                            .type(listenerType)
+                            .retries(String.valueOf(jobRetries))));
+
+    final JobHandler completeJobWithVariableHandler =
+        (jobClient, job) ->
+            jobClient.newCompleteCommand(job.getKey()).variable("my_variable", 123).send().join();
+
+    final var recordingHandler = new RecordingJobHandler(completeJobWithVariableHandler);
+    client.newWorker().jobType(listenerType).handler(recordingHandler).open();
+
+    // when
+    final var userTaskCompletionFuture =
+        client.newUserTaskCompleteCommand(userTaskKey).send().toCompletableFuture();
+    await("until all retries are exhausted")
+        .untilAsserted(
+            () ->
+                assertThat(recordingHandler.getHandledJobs())
+                    .describedAs(
+                        "TL job should retry until reaching the final attempt with retries set to 1")
+                    .last()
+                    .extracting(ActivatedJob::getRetries)
+                    .isEqualTo(1));
+
+    // then
+    final var handledJobs = recordingHandler.getHandledJobs();
+    assertThat(handledJobs)
+        .hasSize(jobRetries)
+        .allSatisfy(job -> assertThat(job.getType()).isEqualTo(listenerType));
+
+    assertThat(handledJobs.getFirst())
+        .describedAs("Job attempts should have same field values except 'retries' and 'deadline'")
+        .usingRecursiveComparison()
+        .ignoringFields("retries", "deadline")
+        .isEqualTo(handledJobs.getLast());
+
+    // assert that an incident was created due to the rejection of TL job completion with variables
+    final var expectedErrorMessageWithRejectionReason =
+        "Command 'COMPLETE' rejected with code 'INVALID_ARGUMENT': Task Listener job completion with variables payload provided is not supported";
+    ZeebeAssertHelper.assertIncidentCreated(
+        incident ->
+            assertThat(incident)
+                .hasJobKey(handledJobs.getLast().getKey())
+                .hasErrorType(ErrorType.JOB_NO_RETRIES)
+                .extracting(
+                    IncidentRecordValue::getErrorMessage, as(InstanceOfAssertFactories.STRING))
+                .startsWith("io.camunda.zeebe.client.api.command.ClientStatusException:")
+                .contains(expectedErrorMessageWithRejectionReason));
+
+    // The rejection of the TL job `COMPLETE` command, due to variables payload being set,
+    // results in the `COMPLETE` user task command request failing with a `timeout` exception.
+    assertThatThrownBy(userTaskCompletionFuture::join)
+        .isInstanceOf(ClientException.class)
+        .hasMessageEndingWith("java.net.SocketTimeoutException: 5 SECONDS");
```

Comment:
> I believe this assumption is incorrect. The retries are not automatically decremented.

As I mentioned in this [comment](https://github.com/camunda/camunda/pull/24182#discussion_r1829330801) this is not the case when the complete job command was sent with `.join()` method to await the completion.
```
jobClient.newCompleteCommand(job.getKey()).variable("bar_var", 123).send().join();
```
If the response contains a rejection, the `ClientStatusException` will be thrown and the job will be retried until all retries are exhausted, so `JOB_NO_RETRIES` incident will be created after.

Reply:
> No, we just don't allow completing a task listener job with variables. Nothing else to do.
> For example, when we reject the first attempt to complete the task listener job with variables, and then see another completion attempt, **but this time without variables**, then we accept it. This will in turn progress the completion of the task and result in a success response to the complete user task request.

👍 I got it.
🙏 Thanks for confirming this.

Reply:
Great idea, @korthout!

With your suggestions, the execution time for this test dropped to ~3.3 from ~10.5 seconds. 

The test has grown quite large. I tried extracting some assertions into utility methods, but they seem too specific to this scenario to generalize effectively. Let’s see what other integration tests we add for Task Listeners, and that might reveal some common patterns to extract.
Or perhaps #23702 will bring variables support back, and this test will fade into Git history :slightly_smiling_face:

Thanks again for the input!